lib/elixir/lib/module/types/descr.ex

```diff
@@ -113,18 +124,36 @@ defmodule Module.Types.Descr do
 
   @compile {:inline, to_quoted: 2}
   defp to_quoted(:bitmap, val), do: bitmap_to_quoted(val)
+  defp to_quoted(:atom, val), do: atom_to_quoted(val)
 
   @doc """
   Converts a descr to its quoted string representation.
   """
   def to_quoted_string(descr) do
     descr
-    |> to_quoted()
+    |> to_quoted_with_print_heuristic()
     |> Code.Formatter.to_algebra()
     |> Inspect.Algebra.format(98)
     |> IO.iodata_to_binary()
   end
 
+  # Heuristic for when to print a type as `not t`: if the number of top types in `t`
+  # is greater than 6, we print it as `not (negation(t))`
+  # Hence, we print `not integer()` instead of `atom() or float() or tuple() or ...`
+  defp to_quoted_with_print_heuristic(descr) do
+    if count_top_types(descr) > 6,
```

Comment:
I have a feeling this heuristic is wrong. For example, if you have `integer() or float() or ... or :foo or :bar`, it will also negate the atoms and we don't want that?

Can we remove the heuristic from this PR for now and we will revisit it later? :) I feel it should probably be something like:

1. If all keys (except bitmap) are a top type AND the number of keys plus bitmaps are more than six, negate
2. If the bitmap has more than 6 types, negate exclusively the bitmap

Reply:
Yes, let's remove it before agreeing on something more precise! 

The heuristic doesn't print wrong types (because `t` is the same as `not (not t)`) so negating `:foo or :bar` here  means we have a double negation, which is more of a problem in understanding the type.

It is not that bad too: in your example the very long
`binary() or empty_list() or integer() or float() or pid() or port() or pid() or reference() or non_empty_list() or map() or tuple() or fun() or :foo or :bar`
gets printed as
`not (atom and not (:foo or :bar))` 
which is hard to read but gives some nice information: "any type except all the atoms that aren't `:foo`, `:bar` "

Since those are all equivalent, I think a better type for printing here would be
`:foo or :bar or not atom()`